src/game/boe.town.cpp

```diff
@@ -1247,52 +1245,79 @@ void erase_specials() {
-							
-						}
-					}
-				}
-				for(int k = 0; k < sector.special_locs.size(); k++) {
-					if(sector.special_locs[k].spec < 0) continue; // TODO: Is this needed? Seems to be important, so why was it commented out?
-					out_num = univ.scenario.outdoors.width() * (univ.party.outdoor_corner.y + j) + univ.party.outdoor_corner.x + i;
-					
-					sn = sector.specials[sector.special_locs[k].spec];
-					sd1 = sn.sd1; sd2 = sn.sd2;
-					if((univ.party.sd_legit(sd1,sd2)) && (PSD[sd1][sd2] == 250)) {
-						where = sector.special_locs[k];
-						if(where.x > 48 || where.y > 48 || where.x < 0 || where.y < 0) {
-							beep();
-							add_string_to_buf("Outdoor section corrupt. Problem fixed.");
-							sector.special_locs[k].spec = -1; // TODO: Again, was there a reason for commenting this out?
-						}
-						
-						sector.special_spot[where.x][where.y] = false;
-					}
-				}
+	for(short i = 0; i < 2; i++) {
+		for (short j = 0; j < 2; j++) {
+			if (!quadrant_legal(i, j)) continue;
+			auto& sector = get_sector_ptr(univ.party.outdoor_corner.x + i, univ.party.outdoor_corner.y + j);
+
+			erase_hidden_towns(sector, i, j);
+
+			erase_completed_specials(sector);
+		}
+	}
+}
+
+void erase_hidden_towns(cOutdoors& sector, int quadrant_x, int quadrant_y)
+{
+	for (short tile_index = 0; tile_index < sector.city_locs.size(); tile_index++) {
+		if (!sector.is_city_on_map(tile_index) ||
+			!is_tile_of_type(sector, tile_index, eTerSpec::TOWN_ENTRANCE) ||
+			!sector.is_tile_on_edge(tile_index)) {
+			continue;
+		}
+		if (!town_exists(sector, tile_index)) { continue; }
+		auto town_pos_x = AREA_MEDIUM * quadrant_x + sector.city_locs[tile_index].x;
+		auto town_pos_y = AREA_MEDIUM * quadrant_y + sector.city_locs[tile_index].y;
+		auto spec_pos_x = sector.city_locs[tile_index].x;
+		auto spec_pos_y = sector.city_locs[tile_index].y;
+		auto area_index = sector.terrain[spec_pos_x][spec_pos_y];
+		if (!univ.scenario.towns[sector.city_locs[tile_index].spec]->can_find) {
+			univ.out[town_pos_x][town_pos_y] = univ.scenario.ter_types[area_index].flag1;
+		} else {
+			univ.out[town_pos_x][town_pos_y] = area_index;
+		}
+	}
+}
+
+void erase_completed_specials(cOutdoors& sector)
+{
+	for (auto tile_index = 0; tile_index < sector.special_locs.size(); tile_index++) {
+		if (sector.special_locs[tile_index].spec < 0) continue; // TODO: Is this needed? Seems to be important, so why was it commented out?
+
+		auto sn = sector.specials[sector.special_locs[tile_index].spec];
+		if (univ.party.sd_legit(sn.sd1, sn.sd2) && PSD[sn.sd1][sn.sd2] == SDF_COMPLETE) {
+			auto where = sector.special_locs[tile_index];
+			if (!is_tile_in_range(where)) {
+				beep();
+				add_string_to_buf("Outdoor section corrupt. Problem fixed.");
+				sector.special_locs[tile_index].spec = -1; // TODO: Again, was there a reason for commenting this out?
 			}
+
+			sector.special_spot[where.x][where.y] = false;
+		}
+	}
+}
+
+cOutdoors& get_sector_ptr(int x, int y)
```

Comment:
This returns a reference, not a pointer, so please name it accordingly - either `get_sector_ref` or simply `get_sector`.

It could also be added to cUniverse or cScenario rather than being a global function, I guess... though maybe that slightly defeats the point?